### ISSUE TITLE
Feat: Auto-open coder_app on avail

### ISSUE THREAD

**Issue by MrPeacockNLB** [coder/coder#6243]
# Description

For currently most of my workspace I open I start code-server after it is available. Thinking broader this can also open up an application started via init script.

Example:

```yaml
resource "coder_app" "unknown" {
  agent_id     = coder_agent.main.id
  slug         = "app8080"
  display_name = "app8080"
  icon         = "/icon/code.svg"
  url          = "http://localhost:8080"
  subdomain    = true
  share        = "public"
  
  auto-open    = window # <== "off", "window", "tab". Defaults to "off"

  healthcheck {
    url       = "http://localhost:8080"
    interval  = 3
    threshold = 10
  }
}
```

The terraform resource should have a new parameter e.G. something like **auto-open**. Possible values *off*, *window*, *tab*.

**Comment by ammario** [comment]
We could let you open an app before it's loaded and show a page with a spinner that refreshes when the health check passes.

**Comment by MrPeacockNLB** [comment]
Please do not close this issue

**Comment by amanibhavam** [comment]
Auto-open with the auto create workspace URL https://github.com/coder/coder/pull/8651 would be a great combo for opening something in Coder without user intervention.

**Comment by michaelbrewer** [comment]
Yes, this would be nice to auto open vscode (from the coder_agent - btw this is confusing). 

But in general will speed up the "open in coder" workflow

**Comment by michaelbrewer** [comment]
@MrPeacockNLB  how about a deep link to open a coder_app (even when the workspace has been stopped?)

This kind of already works for vscode desktop

**Comment by MaximeNoulin** [comment]
@michaelbrewer a deep link wouldn't work for users that only use web UI and doesn't have any soft installed locally, am i right?

In my opinion, it would be perfect to have a link that would start the workspace if stopped, display a loading page while the app is starting, and then redirect to the app once the health check passes (i guess it's what @ammario suggested)

**Comment by michaelbrewer** [comment]
Well a deep link does not just mean opening a native app :). It can also mean that get to want to go in a single click (if possible). 

So ideal experience:


- user has a chrome extension (or via open in coder)
- user clicks on a merge request which targets a specific app within a coder template
- browser redirects to coder which auto-starts a new workspace (or have an option to reuse an existing running workspace)
- once the workspace is running, the git repo is cloned. For a merge requests, it will checkout the relevant branch. 
- the browser then redirects to the coder app. Either a desktop app or web app. 

Deep linking can save users a lot of time. 

- reusing an existing workspace, saves time from starting a new one. 
- git cloning the repo and even the specific branch. 
- opening code editor 
Etc.

**Comment by MaximeNoulin** [comment]
I guess there is several separated issues here:
1- the fact to create a new workspace (either via a chrome extension, open in coder, or other custom workflows)
2- the fact to start a workspace (the git clone or checkout part in your comment)
3- the fact to wait for accessing an app that is currently starting (either for the first time or after an awake) or that has been started in the past but where the workspace is now stopped (and therefore needs to be awake)

Every points are relevant but I guess this issue concerns more the third point.
We are at a point of start where the link can be known for sure, we have :
- the app slug
- the workspace agent slug
- the workspace slug
- the user

Therefore, in my opinion, any access to it should indicate that if the workspace is stopped, it should be started
And if it's already starting, or already started but the given app is still not healthy, to display a loading to wait for it to be healthy

NB: In order to avoid having the user locked because the app is unhealthy (due to an error in status response for instance), I would suggest to have a button or link to allow the user to access the app anyway. I've notice some cases where the app could be considered unhealthy but in fact isn't

**Comment by michaelbrewer** [comment]
Also auto-starting stopped workspaces don't work for chrome apps.

**Comment by bpmct** [comment]
Hey, I created https://github.com/coder/coder/issues/19984 to represent some of the points in this issue. I think it makes a ton of sense to improve the coder_app UX.

I don't think we'll ever have the coder dashboard "pop up" new apps (doesn't feel like an ideal UX) but I think the case where the app URL is visited and its not ready is very important.

**Comment by michaelbrewer** [comment]
Thanks. This is a step in the direction of instant workspaces for teams. 

Many users would like to add a repo to a running workspace. 

And in some cases start with a blank workspace. 

Workflow

- user defines the kind of work that want to do. Java with 32 gig memory, 8 cpu cores and plenty storage and started a workspace 
- user would want to open different git repos from gitlab inside that workspace. 

Only an open in code button the can target that workspace would be amazing.
